### DISCUSSION TITLE
그래들 버전은 6.8.2입니다. 7+ 버전을 사용하시는 경우 읽어주세요.

### DISCUSSION THREAD

**Discussion by fsoftwareengineer** [fsoftwareengineer/todo-application#47] [오류/누락된 부분]
그래들 7+ 버전을 사용하시는 경우 compile 대신 implementation을 사용해 주시기 바랍니다.

빌드 시 에러

```
* Where:
Build file '/Users/alyssakim/IdeaProjects/todo-application/4.3-Spring_Security_Integration/demo/build.gradle' line: 29

* What went wrong:
A problem occurred evaluating root project 'demo'.
> Could not find method compile() for arguments [{group=io.jsonwebtoken, name=jjwt, version=0.9.1}] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.
```

수정 사항
1. jsonwebtoken
수정 전
```
	compile group: 'io.jsonwebtoken', name: 'jjwt', version: '0.9.1'
```
수정 후
```
	implementation group: 'io.jsonwebtoken', name: 'jjwt', version: '0.9.1'
```

2. google guava
수정 전
```
	compile group: 'com.google.guava', name: 'guava', version: '28.1-jre'
```

수정 후
```
	implementation group: 'com.google.guava', name: 'guava', version: '28.1-jre'
```

**Comment by mw7895la** [comment]
Gradle버전이 7+일때  TokenProvider.java 의 create나validateAndGetUserId메소드의 return문 부분 소스코드도 수정되어야 할 것 같은데 확인 부탁드립니다 !

**Reply by fsoftwareengineer** [reply]
그래들 버전과 에러가 나는 부분을 남겨주시면 참고하도록 하겠습니다. 
감사합니다.

**Comment by mw7895la** [comment]
Gradle과 build.gradle 을 사진과 같이 설정 후 실습4-7소스코드를 입력했습니다
책의 내용대로 했을 때 create메소드의 signWith() 는 deprecated ,  
validate 메소드 부분의 parseClaimsJws(token) 입니다.
<img width="502" alt="Gradle_version" src="https://user-images.githubusercontent.com/74583354/148011720-2dcbb548-0935-496a-9b07-64146625f443.PNG">
<img width="632" alt="실습코드4-7" src="https://user-images.githubusercontent.com/74583354/148011723-2dbd0672-5296-4648-9669-8ea7f8b51bab.PNG">
<img width="631" alt="build_gradle" src="https://user-images.githubusercontent.com/74583354/148011725-0e759a1e-ea9a-4865-a48d-bf2214f743b9.PNG">

**Reply by fsoftwareengineer** [reply]
해당 에러와 deprecation의 경우 jsonwebtoken버전이 높아 생기는 문제 같습니다. 따로 [포스트](https://github.com/fsoftwareengineer/todo-application/discussions/48)를 작성했으니 참고 해 주시기 바랍니다.

좋은 질문 해 주셔서 감사합니다. :)

**Reply by gunuduru** [reply]
덧붙여 말씀드리자면, 저도 TokenProvider의 validateAndGetUserId에서 parseClaimsJws에서 에러가 발생했어서,

해결한 코드 공유드립니다.

```
public String validateAndGetUserId(String token) {
        // parseClaimsJws 메서드가 Base64로 디코딩 및 파싱
        // 헤더와 페이로드를 setSigningKey로 넘어온 시크릿을 통해 서명한 후 token의 서명과 비교
        // 위조되지 않았다면 페이로드(Claims) 리턴, 위조라면 예외를 날림
        // 그중 우리는 userId가 필요하므로 getBody를 부른다.
        Claims claims = Jwts.parserBuilder() // Jwts.parser()가 deprecated되어, parserBuilder로 대체
                // 참고: https://velog.io/@guswns3371/develop-feat-auth
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token)
                .getBody();

        return claims.getSubject();
    }
```
